### PR TITLE
Inline operator== and operator!=

### PR DESCRIPTION
**Pull request by ktf**
Avoid triggering -Wnon-template-friend on newer GCC.

### CODE REVIEW

**Review by dwblaikie — APPROVED** [review]
If it does what it says on the tin, sounds good to me.

**Review by kuhar — APPROVED** [review]

**Comment by ktf** [comment]
It does and it breaks windows builds.

**Comment by dyung** [comment]
Hi, we also use gcc internally to build the compiler and were seeing the warnings you are trying to fix here, so I tried applying your fix on our repo and it seems to fail with gcc 9.4.0 as well for the same errors I'm seeing with my Windows build of your fix. Namely this error:
```
/home/dyung/src/git/merge/llvm/include/llvm/ADT/PagedVector.h:220:40: error: ‘size_t llvm::PagedVector<clang::QualType>::Size’ is private within this context
  220 |       assert(LHS.ElementIdx == LHS.PV->Size ||
      |                                ~~~~~~~~^~~~
In file included from /home/dyung/src/git/merge/clang/include/clang/Basic/SourceManager.h:46,
                 from /home/dyung/src/git/merge/clang/lib/Serialization/ASTReader.cpp:61:
/home/dyung/src/git/merge/llvm/include/llvm/ADT/PagedVector.h:46:10: note: declared private here
   46 |   size_t Size = 0;
      |          ^~~~
...
/home/dyung/src/git/merge/llvm/include/llvm/ADT/PagedVector.h:222:23: error: ‘llvm::SmallVector<clang::QualType*, 0> llvm::PagedVector<clang::QualType>::PageToDataPtrs’ is private within this context
  222 |               LHS.PV->PageToDataPtrs[LHS.ElementIdx / PageSize]));
      |               ~~~~~~~~^~~~~~~~~~~~~~
In file included from /home/dyung/src/git/merge/clang/include/clang/Basic/SourceManager.h:46,
                 from /home/dyung/src/git/merge/clang/lib/Serialization/ASTReader.cpp:61:
/home/dyung/src/git/merge/llvm/include/llvm/ADT/PagedVector.h:50:31: note: declared private here
   50 |   mutable SmallVector<T *, 0> PageToDataPtrs;
      |                               ^~~~~~~~~~~~~~
```

**Comment by kuhar** [comment]
So it would seem we would have to make the `operator==` friend of both the vector and the iterator class? Something like:
```c++
template <...>
class PagedVector {
  ...

  class MaterializedIterator;
  friend bool operator==(const MaterializeIterator& lhs, const MaterializedIterator& rhs);

  class MaterializedIterator {
    friend bool operator==(const MaterializeIterator& lhs, const MaterializedIterator& rhs) {
      ...
    }
  };
};

File: llvm/include/llvm/ADT/PagedVector.h [hunk 1]
```diff
@@ -209,36 +209,31 @@ template <typename T, size_t PageSize = 1024 / sizeof(T)> class PagedVector {
...
+      // It should not be possible to build two iterators pointing to non
+      // materialized elements.
+      assert(LHS.ElementIdx == LHS.PV->Size ||
+             (LHS.ElementIdx < LHS.PV->Size &&
+              LHS.PV->PageToDataPtrs[LHS.ElementIdx / PageSize]));
+      assert(RHS.ElementIdx == RHS.PV->Size ||
+             (RHS.ElementIdx < RHS.PV->Size &&
+              RHS.PV->PageToDataPtrs[RHS.ElementIdx / PageSize]));
+      return LHS.ElementIdx == RHS.ElementIdx;
+    }
```
Comment (zygoloid) [inline llvm/include/llvm/ADT/PagedVector.h:227]:
```suggestion
      return LHS.equals(RHS);
    }

  private:
    void verify() {
      assert(ElementIdx == PV->Size ||
             (ElementIdx < PV->Size &&
              PV->PageToDataPtrs[ElementIdx / PageSize]));
    }

    bool equals(const MaterializedIterator &Other) {
      assert(PV == Other.PV);
      verify();
      Other.verify();
      return ElementIdx == Other.ElementIdx;
    }
```

Something like this seems like it should work and avoid the MSVC and GCC errors and the GCC warning.

File: llvm/include/llvm/ADT/PagedVector.h [hunk 2]
```diff
@@ -209,36 +209,31 @@ template <typename T, size_t PageSize = 1024 / sizeof(T)> class PagedVector {
       return PagePtr[ElementIdx % PageSize];
     }
 
+    /// Equality operator.
     friend bool operator==(MaterializedIterator const &LHS,
```
Comment (zygoloid) [inline llvm/include/llvm/ADT/PagedVector.h:213]: While we're changing these: LLVM style is to use `const T &`, not `T const &`.

**Comment by ktf** [comment]
Using @zygoloid suggestion it works on GCC (including 9.4, as tested on godbolt), macOS XCode 15, now waiting for the CI to verify Windows.

@kuhar my understanding of your PR seems to still trigger the -Wnon-template-friend warning on GCC.

Side note: does anyone know how to get LLVM work in godbolt? Using the vspkg package does not work for me (missing includes from llvm/.*).

**Comment by ktf** [comment]
I am waiting for Windows to finish to reformat the PR...

**Comment by ktf** [comment]
Ping @kuhar @zygoloid @dwblaikie. This seems to work without warnings on all the compilers I managed to test (gcc 9.4.0 on godbolt, latest GCC on godbolt, macOS XCode 15, and of course the CI). Adapted from @zygoloid suggestion.

**Review by kuhar — APPROVED** [review]

**Comment by mikaelholmen** [comment]
Hi! Anything preventing pushing this now?

**Comment by ktf** [comment]
I was wondering the same. I do not have write rights.

**Comment by ktf** [comment]
@vgvassilev could you take care of this?

**Pull request merged**